### PR TITLE
ios: support additional options pass from js

### PR DESCRIPTION
**Pull request by zxcpoiu**
Support addition options can pass from js method

#### original:
`updateDisplay = (uuid, displayName, handle)`

#### new:
`updateDisplay = (uuid, displayName, handle, options = null)`
where the optional options could be:
```js
{
  hasVideo: true,
  supportsHolding: true,
  supportsDTMF: true,
  supportsGrouping: true,
  supportsUngrouping: true,
} 
```

#### original:
`displayIncomingCall = (uuid, handle, localizedCallerName, handleType = 'number', hasVideo = false)`

#### new:
`displayIncomingCall = (uuid, handle, localizedCallerName, handleType = 'number', hasVideo = false, options = null)`
where the optional options could be:
```js
{
  supportsHolding: true,
  supportsDTMF: true,
  supportsGrouping: true,
  supportsUngrouping: true,
} 
```

### Note for API compatibility:

1.
Originally, `displayIncomingCall`already has `hasVideo` argument but `updateDisplay` are not. For api compatibility, I did NOT move `displayIncomingCall.hasVideo` inside  `displayIncomingCall.options`

2.
In the earlier time, we've exposed native `reportNewIncomingCall` and introduced an overloading function for the new `payload`argument for api compatibility:

#### Earlier Original (as simple api):
`[RNCallKeep reportNewIncomingCall:uuid handle:handle handleType:@"generic" hasVideo:false localizedCallerName:callerName fromPushKit: YES];`

and 

#### Earlier New (as full api):
`[RNCallKeep reportNewIncomingCall:uuid handle:handle handleType:@"generic" hasVideo:false localizedCallerName:callerName fromPushKit: YES payload:extra];`

We can't have too much overloading functions, so I keep the original one, and extend the new overloading function, so now is:

#### This PR Original (as simple api):
`[RNCallKeep reportNewIncomingCall:uuid handle:handle handleType:@"generic" hasVideo:false localizedCallerName:callerName fromPushKit: YES];`

and 

#### This PR New (as full api):
`[RNCallKeep reportNewIncomingCall:uuid handle:handle handleType:@"generic" hasVideo:false localizedCallerName:callerName supportsHolding: YES supportsDTMF: YES supportsGrouping: YES supportsUngrouping: YES fromPushKit: YES payload: payload];`

### TL;DR

For user using native `RNCallKeep reportNewIncomingCall` with `payload` argument should change to `[RNCallKeep reportNewIncomingCall:uuid handle:handle handleType:@"generic" hasVideo:false localizedCallerName:callerName supportsHolding: YES supportsDTMF: YES supportsGrouping: YES supportsUngrouping: YES fromPushKit: YES payload: payload];`

### CODE REVIEW

**Comment by sboily** [comment]
Hello, does this features are possible on android? I haven't checked.

**Comment by sboily** [comment]
I checked the connection service API and looks like there is nothing to control that. options looks like generic, but what we want to do with android with this options? Does options need to be renamed to be more ios specific? Can you add the documentation on the README?

**Comment by zxcpoiu** [comment]
@sboily 

Yes, on android, currently is no-op to the options.

About the naming, I just left it as it's original name of iOS because in the original api `displayIncomingCall`, there is an arguments called `hasVideo` which is no-op on android as well. 
[see here](https://github.com/react-native-webrtc/react-native-callkeep/blob/master/index.js#L59)

I could change the name in the options with `ios` prefix indicating it's an ios specific option, like: `iosHasVideo`, but it looks like uncoordinated with the original `hasVideo` arg in `displayIncomingCall` described above.

Maybe we could change to like this?

```js
let options = {
  ios: {
    hasVideo: true,
    supportsHolding: true,
    supportsDTMF: true,
    supportsGrouping: true,
    supportsUngrouping: true,
  }
} 
``` 

Because this PR related to api definition and slightly breaking change, so **I will add document once we are all agreed the final api and this PR** then :)

**Comment by sboily** [comment]
Hello, i prefer this approach @zxcpoiu.

**Comment by zxcpoiu** [comment]
@danjenkins @sboily 

To confirm, we are saying that we prefer below approach, right?

```js
let options = {
  ios: {
    hasVideo: true,
    supportsHolding: true,
    supportsDTMF: true,
    supportsGrouping: true,
    supportsUngrouping: true,
  }
} 
```

**Comment by sboily** [comment]
Yes, i also confirm, thank you.

**Comment by manuquentin** [comment]
@zxcpoiu can you rebase on master please ?

**Comment by zxcpoiu** [comment]
Done!

* Changed to platform specific options
* Add documents

If you want the api to be unified, I'm happy to move `hasVideo` arg of `displayIncomingCall` into `options` as well, but this introduced breaking changes to every user which invoke `displayIncomingCall` from js.

Let me know if anything need to change.

**Comment by lrorthmann** [comment]
@zxcpoiu there is a typo in the index.js
```js
...options.ios; // <-- should be a normal comma
```

**Comment by zxcpoiu** [comment]
fixed, thanks @lrorthmann

**Comment by sboily** [comment]
@zxcpoiu is it ready for review and possible merge? CC @manuquentin

**Comment by manuquentin** [comment]
@zxcpoiu can you reabse it please ? So we can merge it.

**Comment by manuquentin** [comment]
@zxcpoiu can you rebase your PR ? Thanks !

**Comment by zxcpoiu** [comment]
sorry for the delay, rebased.
But now seems guthub have some issue, the PR did not refreshed yet :(

**Comment by zxcpoiu** [comment]
But I noticed this PR needs to be discussed about the overloading api `reportNewIncomingCall`

Due to the addition of `withCompletionHandler`, now we have 3 overloading functions for backward compatibility:

### Main functioon

main function should contains all arguments

https://github.com/react-native-webrtc/react-native-callkeep/blob/master/ios/RNCallKeep/RNCallKeep.m#L369

```objc
+ (void)reportNewIncomingCall:(NSString *)uuidString
                       handle:(NSString *)handle
                   handleType:(NSString *)handleType
                     hasVideo:(BOOL)hasVideo
          localizedCallerName:(NSString * _Nullable)localizedCallerName
                  fromPushKit:(BOOL)fromPushKit
                      payload:(NSDictionary * _Nullable)payload
        withCompletionHandler:(void (^_Nullable)(void))completion
{
```
### Overloading 1

When we was adding `payload`, we introduced overloading function for user who is not using `payload`

https://github.com/react-native-webrtc/react-native-callkeep/blob/master/ios/RNCallKeep/RNCallKeep.m#L416

```objc
+ (void)reportNewIncomingCall:(NSString *)uuidString
                       handle:(NSString *)handle
                   handleType:(NSString *)handleType
                     hasVideo:(BOOL)hasVideo
          localizedCallerName:(NSString * _Nullable)localizedCallerName
                  fromPushKit:(BOOL)fromPushKit
{
```

### Overloading 2

When we was adding `withCompletionHandler`, we introduced overloading function for user who is not using `withCompletionHandler` ( but using payload )


https://github.com/react-native-webrtc/react-native-callkeep/blob/master/ios/RNCallKeep/RNCallKeep.m#L358

```objc
+ (void)reportNewIncomingCall:(NSString *)uuidString
                       handle:(NSString *)handle
                   handleType:(NSString *)handleType
                     hasVideo:(BOOL)hasVideo
          localizedCallerName:(NSString * _Nullable)localizedCallerName
                  fromPushKit:(BOOL)fromPushKit
                      payload:(NSDictionary * _Nullable)payload
{
```

**Comment by zxcpoiu** [comment]
### Main function should have all arguments for sure:

```objc
+ (void)reportNewIncomingCall:(NSString *)uuidString
                       handle:(NSString *)handle
                   handleType:(NSString *)handleType
                     hasVideo:(BOOL)hasVideo
          localizedCallerName:(NSString * _Nullable)localizedCallerName
              supportsHolding:(BOOL)supportsHolding
                 supportsDTMF:(BOOL)supportsDTMF
             supportsGrouping:(BOOL)supportsGrouping
           supportsUngrouping:(BOOL)supportsUngrouping
                  fromPushKit:(BOOL)fromPushKit
                      payload:(NSDictionary * _Nullable)payload
        withCompletionHandler:(void (^_Nullable)(void))completion
{
```

### So now  we have to decide:

`what kind` and `how many` overloading functions we should keep for api backward compatibility?

**Comment by sboily** [comment]
Hello @zxcpoiu, yes ! you right we need to cleanup a little that. We could probably release a new major version with some new features like #237 and #188 and removing some old backward compatibilities if there is no strong objection here. I think with these two PRs we have some interesting works to have video capabilities in callkeep for android and iOS.
Can you make a new tentative for updating the PR please?
Thank you!

**Comment by zxcpoiu** [comment]
The commit history is getting funky on github, not sure why. force push and delete branch does't fix it.
I continued at #238
